src/core_init_atmosphere/mpas_init_atm_static.F

```diff
@@ -464,171 +468,240 @@ subroutine init_atm_static(mesh, dims, configs)
-             xPixel = sphere_radius * cos(lon_pt) * cos(lat_pt)    ! at this point, so we need to ues the same radius
-             yPixel = sphere_radius * sin(lon_pt) * cos(lat_pt)    ! for source pixel coordinates
-
-             if (mpas_in_cell(xPixel, yPixel, zPixel, xCell(iPoint), yCell(iPoint), zCell(iPoint), &
-                         nEdgesOnCell(iPoint), verticesOnCell(:,iPoint), xVertex, yVertex, zVertex)) then
-
-                ncat(int(rarray(i,j,1)),iPoint) = ncat(int(rarray(i,j,1)),iPoint) + 1
+     if (all(ncat(:,i) == 0)) then
+         tile => null()
+         ierr = mgr % get_tile(latCell(i), lonCell(i), tile)
+         if (ierr /= 0 .or. .not. associated(tile)) then
+             call mpas_log_write('Could not get the tile that contained cell $i', intArgs=(/i/), messageType=MPAS_LOG_CRIT)
+         end if
 
-             end if
-          end if
-       end do
-       end do
+         ierr = mgr % push_tile(tile)
+         if (ierr /= 0) then
+             call mpas_log_write('Error pushing this tile onto the stack: '//trim(tile%fname), messageType=MPAS_LOG_CRIT)
+         end if
+    end if
+
+    do while(.not. mgr % is_stack_empty())
+        tile => mgr % pop_tile()
+
+        if (tile % is_processed) then
+            cycle
+        end if
+
+        call mpas_log_write('Processing tile: '//trim(tile%fname))
+
+        all_pixels_mapped_to_halo_cells = .true.
+
+        do j = 1 + tile_bdr, tile_ny + tile_bdr, 1
+            do k = 1 + tile_bdr, tile_nx + tile_bdr, 1
+
+                ! TODO: Should we have a check if a tile value is greater or less than
+                ! the category_max or category_min, respectfully?
+
+                call mgr % tile_to_latlon(tile, j, k, lat_pt, lon_pt)
+                call mpas_latlon_to_xyz(x, y, z, sphere_radius, lat_pt, lon_pt)
+
+                res => null()
+                call mpas_kd_search(tree, (/x, y, z/), res)
+
+                !
+                ! For all but the outermost boundary cells, we can safely assume that the nearest
+                ! model grid cell contains the pixel (else, a different cell would be nearest)
+                !
+                if (bdyMaskCell(res % id) < nBdyLayers) then
+
+                   ncat(int(tile % tile(k,j,1)), res % id) = ncat(int(tile % tile(k,j,1)), res % id) + 1
+
+                   if (res % id <= nCellsSolve) then
+                       all_pixels_mapped_to_halo_cells = .false.
+                   end if
+
+                ! For outermost boundary cells, additional work is needed to verify that the pixel
+                ! actually lies within the nearest cell
+                else
+                   ! Model cell coordinates assume a "full" sphere radius
+                   ! at this point, so we need to use the same radius
+                   ! for source pixel coordinates
+                   call mpas_latlon_to_xyz(xPixel, yPixel, zPixel, sphere_radius, lat_pt, lon_pt)
```

Comment:
If we change mpas_kd_search to use `xPixel, yPixel and zPixel` then the above function call and comment can be removed.